lib/src/memory/csr/csr.dart

```diff
@@ -0,0 +1,192 @@
+    // process them individually
+    else {
+      for (final field in config.fields) {
+        if (field.start > currIdx) {
+          fields.add(Logic(
+              name: '${config.name}_rsvd_$rsvdCount',
+              width: field.start - currIdx));
+          rsvds.add(fields.length - 1);
+          rsvdCount++;
+        }
+        fields.add(
+            Logic(name: '${config.name}_${field.name}', width: field.width));
+        currIdx = field.start + field.width;
+      }
+      if (currIdx < config.width) {
+        fields.add(Logic(
+            name: '${config.name}_rsvd_$rsvdCount',
+            width: config.width - currIdx));
+        rsvds.add(fields.length - 1);
+      }
+    }
+    return Csr._(
+      config: config,
+      rsvdIndices: rsvds,
+      fields: fields,
+    );
+  }
+
+  /// Explicit constructor.
+  ///
+  /// This constructor is private and should not be used directly.
+  /// Instead, the factory constructor [Csr] should be used.
+  /// This facilitates proper calling of the super constructor.
+  Csr._({
+    required this.config,
+    required this.rsvdIndices,
+    required List<Logic> fields,
+  }) : super(fields, name: config.name) {
+    config.validate();
+  }
+
+  /// Accessor to the bits of a particular field within the CSR by name [name].
+  Logic getField(String name) =>
+      elements.firstWhere((element) => element.name == '${this.name}_$name');
+
+  /// Accessor to the config of a particular field within the CSR by name [name].
+  CsrFieldConfig getFieldConfigByName(String name) =>
+      config.getFieldByName(name);
+
+  /// Given some arbitrary data [wd] to write to this CSR,
+  /// return the data that should actually be written based
+  /// on the access control of the CSR and its fields.
+  Logic getWriteData(Logic wd) {
+    // if the whole register is ready only, return the current value
+    if (access == CsrAccess.readOnly) {
+      return this;
+    }
+    // register can be written, but still need to look at the fields...
+    else {
+      // special case of no explicit fields defined
+      // in this case, we have an implicit read/write field
+      // so there is nothing special to do
+      if (fields.isEmpty) {
+        return wd;
+      }
+
+      // otherwise, we need to look at the fields
+      var finalWd = wd;
+      var currIdx = 0;
+      var currField = 0;
+      for (var i = 0; i < elements.length; i++) {
+        // if the given field is reserved or read only
+        // take the current value instead of the new value
+        final chk1 = rsvdIndices.contains(i);
+        if (chk1) {
+          finalWd = finalWd.withSet(currIdx, elements[i]);
+          currIdx += elements[i].width;
+          continue;
+        }
+
+        // if the given field is read only
+        // take the current value instead of the new value
+        final chk2 = fields[currField].access == CsrFieldAccess.readOnly ||
+            fields[currField].access == CsrFieldAccess.writeOnesClear;
```

Comment:
Where do you actually control the behavior of "writeOnesClear" doing the "side effect"?